akka-cluster/src/main/scala/akka/cluster/JoinConfigCompatChecker.scala

```diff
@@ -81,7 +107,9 @@ object JoinConfigCompatChecker {
       else Invalid(incompatibleKeys.to[im.Seq])
     }
 
-    exists(requiredKeys, toCheck) ++ checkEquality
+    val checkDiff = diff(requiredKeys, toCheck, actualConfig)
+
+    exists(requiredKeys, toCheck) ++ checkEquality ++ checkDiff
```

Comment:
This became more complicated than I though, and I might still not understand everything.

If I would implement it in a new checker (a new feature for `number-of-shards`) I think everything needed would be:
```
private[akka] final class JoinConfigCompatCheckClusterSharding extends JoinConfigCompatChecker {

  override def requiredKeys: im.Seq[String] =
    im.Seq("akka.cluster.sharding.number-of-shards")

  override def check(toCheck: Config, actualConfig: Config): ConfigValidation = {
    if (toCheck.hasPath("akka.cluster.sharding.number-of-shards"))
      JoinConfigCompatChecker.fullMatch(requiredKeys, toCheck, actualConfig)
    else
      Valid // rolling update, "number-of-shards" doesn't exist in old version
  }
    
}
```

Perhaps we shouldn't try to support it automatically in `fullMatch` but let the concrete checker implementations implement the logic needed to support rolling updates?

Reply:
I started the work with `if (toCheck.hasPath("..")` within the `fullCheck` function and may have found cases that are not so straight forward, but perhaps how I came to see it is incorrect. I'll review.

Reply:
Looking at things again after simplifying and further tests, yes we can not automatically `fullMatch` - it negates too much existing validation logic, and is too risky against expectation.